File: ido_sdk_flutter/ido_sdk/protocol_sdk/protocol_channel/android/src/main/kotlin/com/idosmart/protocol_channel/IDOSDK.kt
package com.idosmart.protocol_channel

import android.content.Context
import android.os.Looper
import com.idosmart.pigeon_implement.*
import com.idosmart.model.*
import com.idosmart.protocol_sdk.*

import io.flutter.embedding.engine.FlutterEngine
import io.flutter.embedding.engine.dart.DartExecutor
import kotlinx.coroutines.*

val sdk: IDOSDK = IDOSDK.instance()

class IDOSDK private constructor() {

    companion object {
        @Volatile
        private var instance: IDOSDK? = null
        @JvmStatic
        fun instance(): IDOSDK {
            return instance ?: synchronized(this) {
                instance ?: IDOSDK().also { instance = it }
            }
        }
    }

    fun init(context: Context) {
        val engine = FlutterEngine(context)
        engine.dartExecutor.executeDartEntrypoint(DartExecutor.DartEntrypoint.createDefault())
    }


    /**
     * 当前协议库操作状态
     */
    @Deprecated("Use sdk.state")
    val protocolDevice: IDOProtocolState
        get() {
            return BridgeDelegateImpl.instance().protocolDevice
        }

    /**
     * 当前协议库操作状态
     */
    val state: IDOProtocolState
        get() {
            return BridgeDelegateImpl.instance().protocolDevice
        }

    //==============do not edit directly begin==============
    /**
     * 蓝牙接口
     */
    val ble: BleInterface by lazy {
        BluetoothImpl()
    }
    //==============do not edit directly end==============

    /**
     * 桥接接口
     */
    val bridge: BridgeInterface by lazy {
        BridgeImpl()
    }

    /**
     * alexa
     */
    val alexa: AlexaInterface by lazy {
        AlexaImpl()
    }

    /**
     * 设备信息
     */
    val device: DeviceInterface by lazy {
        IDODevice()
    }

    /**
     * 功能表
     */
    val funcTable: FuncTableInterface by lazy {
        IDOFuncTable()
    }

    /**
     * 基础指令
     */
    val cmd: CmdInterface by lazy {
        CmdImpl()
    }

    /**
     * 设备日志
     */
    val deviceLog: DeviceLogInterface by lazy {
        DeviceLogImpl()
    }

    /**
     * 消息图标
     */
    val messageIcon: MessageIconInterface by lazy {
        MessageIconImpl()
    }

    /**
     * 数据同步
     */
    val syncData: SyncDataInterface by lazy {
        SyncDataImpl()
    }

    /**
     * 数据交换
     */
    val dataExchange: ExchangeDataInterface by lazy {
        ExchangeDataImpl()
    }

    /**
     * 工具
     */
    val tool: ToolsInterface by lazy {
        ToolImpl()
    }

    /**
     * 文件传输
     */
    val transfer: FileTransferInterface by lazy {
        FileTransferImpl()
    }

    /**
     * sdk info
     */
    val info: IDOSdkInfo
        get() {
            val verInfo = plugin.verInfo
            if (verInfo != null) {
                _sdkInfo.versionLib = verInfo.verMain!!
                _sdkInfo.versionAlexa = verInfo.verAlexa!!
                _sdkInfo.versionClib = verInfo.verClib!!
            }
            return _sdkInfo
        }
    private val _sdkInfo: IDOSdkInfo = IDOSdkInfo()
}

class IDOSdkInfo {
    /**
     * SDK版本
     */
    val versionSdk = "4.5.1"


    /**
     *  SDK更新时间
     */
    val updateTime = "2026-03-26 11:37:52"

    /**
     * Lib库版本
     */
    var versionLib: String = "4.0.16"
        internal set

    /**
     * Alexa库版本
     */
    var versionAlexa: String = "2.0.9"
        internal set

    /**
     * c库版本
     */
    var versionClib: String = "3.7.12"
        internal set
}

internal fun innerRunOnMainThread(closure: () -> Unit) {
    val context = if (isMainThread()) {
        Dispatchers.Main.immediate
    } else {
        Dispatchers.Main
    }

    // 使用协程作用域启动协程，在主线程上运行 closure 而不阻塞线程
    CoroutineScope(context).launch {
        closure()
    }
}

private fun isMainThread(): Boolean {
    return Thread.currentThread() == Looper.getMainLooper().thread
}

File: ido_sdk_flutter/ido_sdk/protocol_sdk/protocol_channel/ios/Classes/protocol_sdk/IDOSDK.swift
//
//  IDOSDK.swift
//  protocol_channel
//
//  Created by hc on 2023/8/15.
//

import Foundation

/// SDK入口类
@objcMembers
public class IDOSDK: NSObject, IDOSdkInterface {
    public static let shared = IDOSDK()
    override private init() {}

    public private(set) lazy var device: IDODeviceInterface = DeviceImpl()

    public private(set) lazy var funcTable: IDOFuncTableInterface = FuncTableImpl()

    public private(set) lazy var cmd: IDOCmdInterface = CmdImpl()

    public private(set) lazy var bridge: IDOBridgeInterface = BridgeImpl()

    public private(set) lazy var ble: IDOBleInterface = BleImpl()

    public private(set) lazy var alexa: IDOAlexaInterface = AlexaImpl()

    public private(set) lazy var transfer: IDOFileTransferInterface = FileTransferImpl()

    public private(set) lazy var messageIcon: IDOMessageIconInterface = MessageIconImpl()

    public private(set) lazy var syncData: IDOSyncDataInterface = SyncDataImpl()
    
    public private(set) lazy var dataExchange: IDODataExchangeOCInterface = ExchangeDataOCImpl()

    public private(set) lazy var deviceLog: IDODeviceLogInterface = DeviceLogImpl()

    public private(set) lazy var tool: IDOToolsInterface = ToolImpl()
    
    public private(set) lazy var state: IDOProtocolState = BridgeDelegateImpl.shared.protocolState
    
    private let _sdkInfo = IDOSdkInfo()
    public var info: IDOSdkInfo {
        get {
            guard let verInfo = SwiftProtocolChannelPlugin.shared.verInfo else {
                return _sdkInfo
            }
            _sdkInfo.versionLib = verInfo.verMain!
            _sdkInfo.versionAlexa = verInfo.verAlexa!
            _sdkInfo.versionClib = verInfo.verClib!
            return _sdkInfo
        }
    }
}


@objcMembers
public class IDOSdkInfo: NSObject {
    /// SDK版本
    public let versionSdk = "4.5.1"
    
    /// SDK更新时间
    public let updateTime = "2026-03-26 11:37:52"
    
    /// Lib库版本
    public internal(set) var versionLib: String = "4.0.16"
    
    /// Alexa库版本
    public internal(set) var versionAlexa: String = "2.0.9"
    
    /// c库版本
    public internal(set) var versionClib: String = "3.7.12"
}

func _runOnMainThread(_ closure: @escaping () -> Void) {
    if Thread.isMainThread {
        closure()
    } else {
        DispatchQueue.main.async {
            _logNative("其它线程，切主线程后调用")
            closure()
        }
    }
}
